### ISSUE TITLE
Smarter unwrapping for `Optional` type possible?

### ISSUE THREAD

**Issue by norman-thomas** [python/mypy#4245]
Hi,
First of all: Thank you all for this great tool!

I am currently using mypy 0.540 and came across an inconvenience when using the `Optional` type.
mypy seems smart enough to detect that after checking an optional value for `None` the `Optional` part of the type is disregarded. Meaning: declaring `value: Optional[int]` would result in `value` behaving like an `int` after `if value is not None`.

So far, so good. If that `value is not None` check is saved into a variable before, the check fails and mypy complains about `Optional[int]` not being compatible with `int`.

I attached a small demo script for this scenario:

```python
from typing import Optional

def strict_function(param: int) -> None:
    print(param)

value: Optional[int] = None

has_value = value is not None
if has_value:
    reveal_type(value)  # Revealed type is 'Union[builtins.int, builtins.None]'
    strict_function(value)  # causes error

if value is not None:
    reveal_type(value)  # Revealed type is 'builtins.int'
    strict_function(value)  # doesn't cause error
```

After invoking `mypy --strict-optional test.py` I get:
```
test.py:10: error: Argument 1 to "strict_function" has incompatible type "Optional[int]"; expected "int"
```

I admit, I sometimes save those checks in `boolean` variables for better code readability as well as avoiding repetitive checks.

It would be great if mypy could also detect these scenarios. I'm not expecting more complex checks to work, but simple ones like in this demo script would be great.

**Comment by gvanrossum** [comment]
This is occasionally requested (there's probably already an issue) but it's a much harder thing to implement, so it probably won't happen any time soon.

**Comment by manoadamro** [comment]
I know this is old but for anyone stuck on this, the only way around this I could find is to redefine the variable after the null check.
```
x: Optional[int] = maybe()
if x is None:
    raise Exception()

actual_x: int = x
```

If anyone has a better solution, let me know

**Comment by kyprifog** [comment]
I've been doing the same thing, this may be the best you can hope for in python.  mypy will still check the unwrapped type, but this depends on erasure.

**Comment by Akuli** [comment]
`assert` can be useful for this:

```python3
x: Optional[int] = maybe()
assert x is not None
actual_x: int = x
```

**Comment by alairock** [comment]
> `assert` can be useful for this:
> 
> ```python
> x: Optional[int] = maybe()
> assert x is not None
> actual_x: int = x
> ```

Careful though, https://docs.python.org/3/reference/simple_stmts.html#assert

`Assert` will not always be available in a production environment. In a test environment, like pytest, using assert is fine.

**Comment by Akuli** [comment]
I know that asserts can be disabled, and even with disabled asserts, trying to use None will likely cause an exception anyway. However, the assert makes mypy accept the code.

**Comment by kyprifog** [comment]
Are you guys familiar with this library?  https://github.com/dry-python/returns

**Comment by mschoettle** [comment]
> ```python
> x: Optional[int] = maybe()
> if x is None:
>     raise Exception()
> 
> actual_x: int = x
> ```

Thanks! It didn't work for me with calling `maybe()` directly in the if statement but additionally declaring the variable works.

**Comment by Ashesh3** [comment]
> This is occasionally requested (there's probably already an issue) but it's a much harder thing to implement, so it probably won't happen any time soon.

I think i am stuck on a similar issue but with Optional int in List.

The following demo snippet throws an error:

```
a: List[Optional[int]] = [None, None]

a[0] = 5
a[1] = 10

b: int = 0

if a[0] != None:
    b = a[0]
```
```
test.py:12: error: Incompatible types in assignment (expression has 
type "Optional[int]", variable has type "int")
Found 1 error in 1 file (checked 1 source file)
```

I do check if `a[0] is not None` but i don't know why it still fails to narrow down the type of a[0] from `int | None` to `int` as the condition for `None` is already checked for.

**Comment by dubois** [comment]
Here is an alternative to `cast` that I have used successfully.
```
U = TypeVar('U')
def nn(inst: Optional[U]) -> U:
  """Not-none helper"""
  assert inst is not None
  return inst
```
used like so
```
    def do_something(a: int): pass

    if node.type == Node.TYPE_Int:
      # node.int's type is Optional[int] because it's filled in if and only if type == TYPE_Int
      do_something(node.int)      # error: ... type "Optional[int]"; expected "int"
      do_something(nn(node.int))  # okay
      node.int= nn(node.int)
      do_something(node.int)      # also okay
```

**Comment by kopecn** [comment]
not pretty but -->
```
def getNested(object, attributes: list[str]):
    recurse = object
    for item in attributes:
        recurse = getattr(recurse, item, None)
    return recurse
```

**Comment by JelleZijlstra** [comment]
The original request here is essentially the same as #11979. This issue is much older but is much less focused, so closing it.
